### PR TITLE
Second round of refactoring to get rendering working.

### PR DESCRIPTION
**Pull request by mdbenjam**
Connects: #3251 

This is a pretty exciting PR. It greatly simplifies how we render pages. We do this by moving as much logic as possible into the PdfPage component. I did a lot of testing to get to this change. Some important things I noticed:

1) Rendering text doesn't take that long.
1) Updating nested redux state can be very costly.
1) If we limit the number of pages we want to render to ~10, then doing them in parallel is pretty fast.

Changes to Pdf.jsx:
1) Keep track of the center of the scroll window component, as well as how much it's scrolled.
1) Remove the functions that used to determine which pages should be drawn. This logic is now in PdfPage.

Changes to PdfPage.jsx:
1) Decouple drawing text from rendering pages.
1) Do more processing on mounting. Get the text, dimensions, and page since these can be costly operations, and once you do them once, we don't need to do them again. Save all these values in the redux state.
1) Choose when to render a page by looking at its distance from the center of the scrollWindow. We set an arbitrary distance and pages within it are rendered. Outside they are not.

Changes to Redux store:
1) Flatten the highly nested documentsByFile object to a pages object.

Things to still fix:
1) Prerendering is broken. Moving to pdfs you haven't visited yet is slow. I'll be tackling this in the next PR.

**Test Plan** 
- [ ] Play around with the pdfs and make sure loading is relatively smooth.

### CODE REVIEW

File: client/app/reader/Pdf.jsx [hunk 1]
```diff
@@ -110,6 +112,10 @@ export class Pdf extends React.PureComponent {
   }
 
   scrollEvent = () => {
+    this.setState({
```
Comment (mdbenjam) [inline client/app/reader/Pdf.jsx]: We use the scrollTop state to signal to the pdf page components that what is in view may have changed.

File: client/app/reader/Pdf.jsx [hunk 2]
```diff
@@ -135,46 +141,6 @@ export class Pdf extends React.PureComponent {
     if (this.props.jumpToPageNumber) {
       this.props.resetJumpToPage();
     }
-    this.drawInViewPages();
-  }
-
-  drawInViewPages = () => {
```
Comment (mdbenjam) [inline client/app/reader/Pdf.jsx]: We can get rid of this complicated method. Each page is now in charge of when it should render.

File: client/app/reader/Pdf.jsx [hunk 3]
```diff
@@ -452,51 +416,6 @@ export class Pdf extends React.PureComponent {
     /* eslint-enable no-negated-condition */
   }
 
-  preDrawPages = () => {
```
Comment (mdbenjam) [inline client/app/reader/Pdf.jsx]: We can get rid of this complicated method. Each page is now in charge of when it should render.

File: client/app/reader/PdfPage.jsx [hunk 4]
```diff
@@ -50,117 +52,147 @@ export class PdfPage extends React.Component {
...
     }
+    this.previousShouldDraw = shouldDraw;
   }
 
-  getDimensions = () => {
-    this.props.pdfDocument.getPage(pageNumberOfPageIndex(this.props.pageIndex)).then((pdfPage) => {
-      const PAGE_DIMENSION_SCALE = 1;
-      const viewport = pdfPage.getViewport(PAGE_DIMENSION_SCALE);
-      const pageDimensions = _.pick(viewport, ['width', 'height']);
+  drawText = () => {
```
Comment (tejans24) [inline client/app/reader/PdfPage.jsx]: Am I correct to assume that drawText is an independent operation from drawPage?

File: client/app/reader/PdfPage.jsx [hunk 5]
```diff
@@ -50,117 +52,147 @@ export class PdfPage extends React.Component {
...
+  }
 
-      // If the scale has changed, draw the page again at the latest scale.
-      if (currentScale !== this.props.scale) {
-        return this.drawPage();
-      }
+  clearPage = () => {
+    if (this.isDrawn) {
+      this.canvas.getContext('2d', { alpha: false }).clearRect(0, 0, this.canvas.width, this.canvas.height);
+      this.props.page.cleanup();
```
[the remark this answers is not in the document]
Comment (mdbenjam) [reply client/app/reader/PdfPage.jsx]: part of pdfjs.

File: client/app/reader/PdfPage.jsx [hunk 6]
```diff
@@ -50,117 +52,147 @@ export class PdfPage extends React.Component {
...
+    const shouldDraw = this.shouldDrawPage(this.props);
 
-    if (this.props.shouldDraw) {
-      if (!prevProps.shouldDraw || prevProps.scale !== this.props.scale) {
-        drawAndUpdateState();
+    // We draw the page if there's been a change in the 'shouldDraw' state, scale, or if
+    // the page was just loaded.
+    if (shouldDraw) {
+      if (this.props.page) {
+        if (!this.previousShouldDraw || prevProps.scale !== this.props.scale || !prevProps.page) {
```
[the remark this answers is not in the document]
Comment (mdbenjam) [reply client/app/reader/PdfPage.jsx]: We need to check if there's been some transition that means we need to draw this now. One of those transitions is we didn't need to draw this before, and now we do.

File: client/app/reader/PdfPage.jsx (hunk 4 above)
Comment (mdbenjam) [reply client/app/reader/PdfPage.jsx]: yes

File: client/app/reader/PdfPage.jsx [hunk 7]
```diff
@@ -50,117 +52,147 @@ export class PdfPage extends React.Component {
...
+    PDFJS.renderTextLayer({
+      textContent: this.props.text,
+      container: this.textLayer,
+      viewport,
+      textDivs: []
     });
   }
 
+  getText = (page) => {
+    return page.getTextContent();
```
[the remark this answers is not in the document]
Comment (mdbenjam) [reply client/app/reader/PdfPage.jsx]: good catch.

**Comment by tejans24** [comment]
Notes: 
* I still see black pages before the pages are rendered.
* Pages don't render unless they are scrolled to
* I noticed the text layer is rendered before pages, thus allowing the text to be searched; which is awesome. But during the search, the page takes a second or two before it's rendered.

I am guessing these 'issues' will be taken care of in the future PRs.

File: client/app/reader/Pdf.jsx [hunk 8]
```diff
@@ -378,7 +344,6 @@ export class Pdf extends React.PureComponent {
 
   componentDidMount() {
     PDFJS.workerSrc = this.props.pdfWorker;
-    window.addEventListener('resize', this.drawInViewPages);
```
[the remark this answers is not in the document]
Comment (mdbenjam) [reply client/app/reader/Pdf.jsx]: I was thinking we didn't need it. But upon reconsideration I think we need to redefine it.

File: client/app/reader/PdfPage.jsx [hunk 9]
```diff
@@ -50,117 +52,147 @@ export class PdfPage extends React.Component {
...
+    }
+
+    const shouldDraw = this.shouldDrawPage(this.props);
 
-    if (this.props.shouldDraw) {
-      if (!prevProps.shouldDraw || prevProps.scale !== this.props.scale) {
-        drawAndUpdateState();
+    // We draw the page if there's been a change in the 'shouldDraw' state, scale, or if
+    // the page was just loaded.
+    if (shouldDraw) {
```
[the remark this answers is not in the document]
Comment (mdbenjam) [reply client/app/reader/PdfPage.jsx]: I combined the 2nd and 3rd ifs, but the first has an attached else, so it needs to stay separate.

**Comment by tejans24** [comment]
bug

I noticed that some documents randomly don't render while flipping through documents.

**Comment by mdbenjam** [comment]
I added your comment on "I noticed that some documents randomly don't render while flipping through documents." to the AC of the ticket. I think it makes sense to deal with in the next PR since that is where we will split up the Pdf component into several PdfFile components.

As for the others. I'll make un-rendered pages white, or maybe we can put on a loading spinner. But, with these changes, we will not be drawing the entire PDF at any given moment. So you might encounter non-rendered pages when scrolling. You just should never encounter them when scrolling at a normal rate.

**Pull request merged**